### PR TITLE
NetworkChaos: fix `both` direction has no effect

### LINKED ISSUE #1486: NetworkChaos direction:Both has no affect
## Bug Report

**What version of Kubernetes are you using?**
```
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.1", GitCommit:"206bcadf021e76c27513500ca24182692aabd17e", GitTreeState:"clean", BuildDate:"2020-09-09T11:18:22Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
```

**What version of Chaos Mesh are you using?**
```
Controller manager Version: version.Info{GitVersion:"v1.1.1", GitCommit:"88e265c1ad2bcc3931645371a71683dd52b576fb", BuildDate:"2021-01-21T11:24:39Z", GoVersion:"go1.14.6", Compiler:"gc", Platform:"linux/amd64"}
```

**What did you do?**
We have the following `NetworkChaos` experiment spec and two pods of the target application running, but when running there appears no network loss at all:

```
spec:
  action: loss
  mode: all
  value: ''
  selector:
    namespaces:
      - <app_namespace>
    labelSelectors:
      app: <app_label_name>
  loss:
    loss: '100'
    correlation: '100'
  direction: both
```

Testing using separate `to` and `from` NetworkChaos experiment specs does appear to work for us.  The configuration used for these specs is as follows:

**From:**
```
spec:
  action: loss
  direction: from
  loss:
    correlation: "100"
    loss: "100"
  mode: all 
  selector:
    namespaces:
    - <app_namespace>
  target:
    mode: all 
    selector:
      labelSelectors:
        app: <app_label_name>
      namespaces:
      - <app_namespace>
    value: ""
  value: ""
```

In this we are limiting (non-target section `selector` option) on namespace as we wish to affect traffic from all pods in that given namespace going to our target pod.

**To:**
```
spec:
  action: loss
  direction: to
  loss:
    correlation: "100"
    loss: "100"
  mode: all 
  selector:
    labelSelectors:
      app: <app_label_name>
    namespaces:
    - <app_namespace>
  target:
    mode: all 
    selector:
      namespaces:
      - <app_namespace>
    value: ""
  value: ""

```

**What did you expect to see?**
Bi-directional network loss for pods 

**What did you see instead?**
No network loss in either direction for the target pods 

**Output of chaosctl**

Not captured

-----
Please advise if I am doing something wrong, but I could not get anything working using the [examples](https://github.com/chaos-mesh/chaos-mesh/blob/master/examples/network-loss-example.yaml) or [documentation](https://chaos-mesh.org/docs/chaos_experiments/networkchaos_experiment).

### PR DESCRIPTION
**Pull request by cwen0**
### What problem does this PR solve?

Fix #1486 #1450 
### What is changed and how does it work?

If the target and external targets are both empty, we don't need to set `dstIpset`. 

### Checklist

Tests

- [ ] Unit test
- [ ] E2E test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

image

### CODE REVIEW

**Comment by YangKeao** [comment]
Setting direction to `To` (as the default value) if the target is empty seems to be a better idea. (because `from` will also cause a misleading behavior).

**Comment by cwen0** [comment]
> Setting direction to `To` (as the default value) if the target is empty seems to be a better idea. (because `from` will also cause a misleading behavior).

If users set the direction to `both`, this action will change the definition of chaos experiment. I don't think this is not good behavior. About `from` will also cause a misleading behavior`, we should verify the chaos experiment when to apply it.

**Comment by cwen0** [comment]
> > Setting direction to `To` (as the default value) if the target is empty seems to be a better idea. (because `from` will also cause a misleading behavior).
> 
> If users set the direction to `both`, this action will change the definition of chaos experiment. This behavior is not friendly for users. About `from will also cause a misleading behavior`, we should verify the chaos experiment when to apply it.

I will update the validatewebhook  to check it.

File: controllers/networkchaos/trafficcontrol/types.go [hunk 1]
```diff
@@ -128,7 +128,10 @@ func (r *endpoint) Apply(ctx context.Context, req ctrl.Request, chaos v1alpha1.I
 			return err
 		}
 	case v1alpha1.Both:
-		err = r.applyTc(ctx, pods, pods, externalCidrs, m, networkchaos)
+		if len(targets) > 0 {
+			targets = pods
+		}
+		err = r.applyTc(ctx, pods, targets, externalCidrs, m, networkchaos)
```
Comment (STRRL) [inline controllers/networkchaos/trafficcontrol/types.go:134]: I have a question, I see that we resolve the "direction: both" with certain codes, it also bring another issue: https://github.com/chaos-mesh/chaos-mesh/issues/1450

Why not this:

```go
case v1alpha1.Both:
    err = r.applyTc(ctx, sources, targets, externalCidrs, m, networkchaos)
    // resolve error
    err = r.applyTc(ctx, targets, sources, []string{}, m, networkchaos)
    // resolve error
```

It means "direction: both" is the composition of both "direction: from" and "direction: to".

**Review by STRRL — COMMENTED** [review]
PTAL @cwen0

File: controllers/networkchaos/trafficcontrol/types.go (hunk 1 above)
Comment (cwen0) [reply controllers/networkchaos/trafficcontrol/types.go:134]: Great! I will fix it.

File: api/v1alpha1/networkchaos_webhook.go [hunk 2]
```diff
@@ -138,16 +138,22 @@ func (in *NetworkChaos) ValidatePodMode(spec *field.Path) field.ErrorList {
...
 			field.Invalid(target.Child("direction"), in.Spec.Direction,
 				fmt.Sprintf("external targets cannot be used with `from` direction in netem action yet")))
 	}
 
+	if in.Spec.Direction == From && in.Spec.ExternalTargets == nil && in.Spec.Target == nil {
+		allErrs = append(allErrs,
+			field.Invalid(target.Child("direction"), in.Spec.Direction,
+				fmt.Sprintf("`from` direction cannot be used when targets and external targets are both empty")))
+	}
+
```
[the remark this answers is not in the document]
Comment (cwen0) [reply api/v1alpha1/networkchaos_webhook.go:156]: This check is ok! I will update it.

**Review by STRRL — DISMISSED** [review]
LGTM, and this PR could fix #1450. I will append e2e test cases soon.

File: controllers/networkchaos/trafficcontrol/types.go (hunk 1 above)
Comment (YangKeao) [reply controllers/networkchaos/trafficcontrol/types.go]: This has another problem :crying_cat_face: . #1448 assumes that the ipset with the same name should have the same content. However, if you `applyTc` twice, they will have the same `ipset` name with different content.

```
dstIpset := ipset.BuildIPSet(targets, externalTargets, networkchaos, string(tcType)[0:5], m.Source)
```

The `postFix` should be modified from `string(tcType)[0:5]` to the `tgt`, `src` like the partition chaos

File: controllers/networkchaos/trafficcontrol/types.go [hunk 3]
```diff
@@ -128,9 +128,15 @@ func (r *endpoint) Apply(ctx context.Context, req ctrl.Request, chaos v1alpha1.I
 			return err
 		}
 	case v1alpha1.Both:
-		err = r.applyTc(ctx, pods, pods, externalCidrs, m, networkchaos)
+		err = r.applyTc(ctx, sources, targets, externalCidrs, m, networkchaos)
```
Comment (YangKeao) [inline controllers/networkchaos/trafficcontrol/types.go:131]: There is another problem :crying_cat_face: . #1448 assumes that the ipset with the same name should have the same content. However, if you `applyTc` twice, they will have the same `ipset` name with different content.

```
dstIpset := ipset.BuildIPSet(targets, externalTargets, networkchaos, string(tcType)[0:5], m.Source)
```

The `postFix` should be modified from `string(tcType)[0:5]` to the `tgt`, `src` like the partition chaos

**Review by YangKeao — CHANGES_REQUESTED** [review]

File: controllers/networkchaos/trafficcontrol/types.go (hunk 3 above)
Comment (cwen0) [reply controllers/networkchaos/trafficcontrol/types.go]: Fixed

Comment (YangKeao) [reply controllers/networkchaos/trafficcontrol/types.go]: Why `toTargetIPSetPostFix` and `bothTargetIPSetPostFix` are different? Should we use the same pattern for these names?

File: controllers/networkchaos/trafficcontrol/types.go [hunk 4]
```diff
@@ -283,8 +301,12 @@ func (r *endpoint) applyTc(ctx context.Context, sources, targets []v1.Pod, exter
 		return nil
 	}
 
+	if len(ipSetPostFix) == 0 {
+		ipSetPostFix = string(tcType[0:5])
+	}
+
 	// create ipset contains all target ips
-	dstIpset := ipset.BuildIPSet(targets, externalTargets, networkchaos, string(tcType)[0:5], m.Source)
+	dstIpset := ipset.BuildIPSet(targets, externalTargets, networkchaos, ipSetPostFix, m.Source)
```
Comment (STRRL) [inline controllers/networkchaos/trafficcontrol/types.go:309]: How about that:

```go
dstIpset := ipset.BuildIPSet(targets, externalTargets, networkchaos, string(tcType)[0:5] + ipSetPostFix, m.Source)
```

It avoids the modification for `ipSetPostFix`, and still shows the `tcType` in the ipset name whether `ipSetPostFix` is empty or not.

Comment (YangKeao) [reply controllers/networkchaos/trafficcontrol/types.go:309]: Maybe we could also shrink the length of `string(tcType)[0:5]`, as I know the only two possible `TcType` is `netem` and  `bandwidth`, and they are easy to distinguish

Comment (cwen0) [reply controllers/networkchaos/trafficcontrol/types.go:309]: `string(tcType)[0:2]` ?

**Review by STRRL — APPROVED** [review]
LGTM

**Review by YangKeao — APPROVED** [review]
LGTM

**Pull request merged**